alert if compactor runs out of disk space (#8278)
grafana/mimir: docs/sources/mimir/manage/mimir-runbooks/_index.md

      - If you encounter any Compactor resource issues, add CPU/Memory as needed temporarily, then scale back later.
      - You can also optionally scale replicas and shards further to split the work up into even smaller pieces until the situation has recovered.

### MimirCompactorHasRunOutOfDiskSpace

This alert fires when the compactor has run out of disk space at least once.
When this happens the compaction will fail and after some time the compactor will retry the failed compaction.

> Unless the compactor concurrency is >=1 it is very likely that on each retry it will just hit the same disk space limit again and it won't be able to recover on its own.

Reviewer: I would perhaps try to explain it:

```suggestion
It is very likely that on each retry of the job, compactor will just hit the same disk space limit again and it won't be able to recover on its own. If compactor concurrency is higher than 1, then it could have been just an unlucky combination of jobs that caused compactor to run out of disk space.
```

> It is very likely that on each retry of the job, compactor will just hit the same disk space limit again and it won't be able to recover on its own. If compactor concurrency is higher than 1, then it could have been just an unlucky combination of jobs that caused compactor to run out of disk space.

Reviewer 2: ```suggestion
It's very likely that on each retry of the job,  the compactor will just hit the same disk space limit again and it won't be able to recover on its own.
Alternatively, if compactor concurrency is higher than 1, it could have been just an unlucky combination of jobs that caused compactor to run out of disk space.
```

How to **investigate**:

- Look at the disk space usage in the compactor's data volumes.

> - Look for an error with the string "no space left on device" to confirm that the compactor ran out of disk space.

Reviewer 2: https://grafana.com/docs/writers-toolkit/write/shortcodes/#code

```suggestion
- Look for an error with the string `no space left on device` to confirm that the compactor ran out of disk space.
```

How to **fix** it:

- The only long-term solution is to give the compactor more disk space, as it requires more space to fit the largest single job into its disk.

> - If the number of blocks that the compactor is failing to compact is not very significant and you want it to skip them in order to focus on more recent blocks instead, you can consider marking the affected blocks for no compaction:

Reviewer 2: ```suggestion
- If the number of blocks that the compactor is failing to compact is not very significant and you want to skip compacting them and focus on more recent blocks instead, consider marking the affected blocks for no compaction:
```

      - If you encounter any Compactor resource issues, add CPU/Memory as needed temporarily, then scale back later.
      - You can also optionally scale replicas and shards further to split the work up into even smaller pieces until the situation has recovered.

### MimirCompactorHasRunOutOfDiskSpace

This alert fires when the compactor has run out of disk space at least once.

> When this happens the compaction will fail and after some time the compactor will retry the failed compaction, but it is very likely that on each retry it will just hit the same disk space limit again and it won't be able to recover on its own.

Reviewer: This is true when using compactor concurrency = 1. With higher value, it's all about what combination of jobs is running at the same time, and it's much trickier to argue that same condition happens again.

This is also reason why we prefer to use concurrency=1 and scale compactors horizontally.

Author: How about this: https://github.com/grafana/mimir/pull/8278/commits/8bb0ed1a03c6a42cdcacd45c5d446c1b1c1adb8d

How to **investigate**:

- Look at the disk space usage in the compactor's data volumes.
- Look for an error with the string "no space left on device" to confirm that the compactor ran out of disk space.

How to **fix** it:

- The only long-term solution is to give the compactor more disk space, this can be achieved by doing one or multiple of:

>   - Horizontally scale the compactors

Reviewer: If there was one job that used more disk space than compactors have, how does adding more compactor replicas help?

Author: Updated in this commit: https://github.com/grafana/mimir/pull/8278/commits/8bb0ed1a03c6a42cdcacd45c5d446c1b1c1adb8d

  - Grow each compactor's disk
- If the number of blocks that the compactor is failing to compact is not very significant and you want it to skip them in order to focus on more recent blocks instead, you can consider marking the affected blocks for no compaction:
  ```

>   # in this example the object storage backend is GCS

Reviewer 2: We discourage comments in the code blocks and prefer the explanation to be in the text preceding the block.

Author: removed in this commit, it's pretty obvious that this is GCS anyway with the `-backend gcs` so I don't think this really needs an explanation:

https://github.com/grafana/mimir/pull/8278/commits/8bb0ed1a03c6a42cdcacd45c5d446c1b1c1adb8d
